File: src/domain/ids.ts
/**
 * Time-sortable identifier generation.
 *
 * Records use UUIDv7-style ids: the leading bits encode a millisecond
 * timestamp, so ids sort chronologically while remaining unique and
 * generatable entirely offline. `crypto.randomUUID` is v4 (not sortable), so a
 * small spec-shaped v7 implementation is provided here.
 */
export function newId(): string {
  const unixMs = Date.now();
  const bytes = new Uint8Array(16);

  // 48-bit big-endian millisecond timestamp in the first 6 bytes.
  bytes[0] = (unixMs / 2 ** 40) & 0xff;
  bytes[1] = (unixMs / 2 ** 32) & 0xff;
  bytes[2] = (unixMs / 2 ** 24) & 0xff;
  bytes[3] = (unixMs / 2 ** 16) & 0xff;
  bytes[4] = (unixMs / 2 ** 8) & 0xff;
  bytes[5] = unixMs & 0xff;

  // Remaining 10 bytes are random.
  crypto.getRandomValues(bytes.subarray(6));

  // Set version (7) and RFC 4122 variant bits.
  bytes[6] = (bytes[6] & 0x0f) | 0x70;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;

  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

File: legacy-python/src/stepbylearn/domain/ids.py
"""Time-sortable identifier generation.

Aggregate roots use UUIDv7-style identifiers: the leading bits encode a
millisecond timestamp, so IDs sort chronologically while remaining globally
unique and generatable entirely offline (no database round-trip). Python's
standard library lacks a UUIDv7 constructor before 3.14, so a small, spec-shaped
implementation is provided here.
"""

from __future__ import annotations

import os
import time
import uuid


def new_id() -> str:
    """Generate a UUIDv7-style, lexicographically sortable identifier.

    The layout follows RFC 9562 §5.7: a 48-bit Unix-millisecond timestamp,
    the version nibble (7), 12 bits of randomness, the variant bits, and 62
    further random bits.

    Returns:
        The identifier as a canonical 36-character UUID string.
    """
    unix_ms = time.time_ns() // 1_000_000
    # 10 random bytes fill everything after the 48-bit timestamp.
    rand = os.urandom(10)

    # Assemble the 128-bit value: 48-bit timestamp in the most-significant bits.
    value = unix_ms << 80
    value |= rand[0] << 72
    value |= rand[1] << 64
    value |= int.from_bytes(rand[2:], "big")

    # Force the version (7) and RFC 4122 variant bits into place.
    value &= ~(0xF000 << 64)  # clear version nibble
    value |= 0x7000 << 64  # set version 7
    value &= ~(0xC000 << 48)  # clear variant bits
    value |= 0x8000 << 48  # set variant 0b10

    return str(uuid.UUID(int=value))
